File: lib/engine/effects.dart
import 'dart:math' as math;
import 'dart:ui';

class Bubble {
  Bubble(this.pos, this.radius, this.speed, this.wobble, this.phase);
  Offset pos;
  double radius;
  double speed;
  double wobble;
  double phase;
  double life = 1;
}

class Droplet {
  Droplet(this.pos, this.vel, this.radius);
  Offset pos;
  Offset vel;
  double radius;
  double life = 1;
}

/// Purely cosmetic particle systems: fizzy bubbles inside the cola and
/// splash droplets at the impact point.
class Effects {
  Effects([int seed = 7]) : _rng = math.Random(seed);

  final math.Random _rng;
  final List<Bubble> bubbles = <Bubble>[];
  final List<Droplet> droplets = <Droplet>[];

  double _bubbleAcc = 0;

  void clear() {
    bubbles.clear();
    droplets.clear();
  }

  void spawnBubbles({
    required double dt,
    required double rate,
    required double left,
    required double right,
    required double bottomY,
    required double surfaceY,
  }) {
    if (right <= left || bottomY <= surfaceY) return;
    _bubbleAcc += rate * dt;
    while (_bubbleAcc >= 1) {
      _bubbleAcc -= 1;
      final x = lerp(left, right, _rng.nextDouble());
      final y = lerp(surfaceY, bottomY, 0.35 + _rng.nextDouble() * 0.65);
      bubbles.add(Bubble(
        Offset(x, y),
        1.2 + _rng.nextDouble() * 2.6,
        26 + _rng.nextDouble() * 52,
        3 + _rng.nextDouble() * 7,
        _rng.nextDouble() * math.pi * 2,
      ));
    }
    if (bubbles.length > 220) {
      bubbles.removeRange(0, bubbles.length - 220);
    }
  }

  void splash(Offset at, double strength) {
    final n = (strength * 3).round().clamp(0, 4);
    for (int i = 0; i < n; i++) {
      final a = -math.pi / 2 + (_rng.nextDouble() - 0.5) * 2.2;
      final sp = 60 + _rng.nextDouble() * 180 * strength;
      droplets.add(Droplet(
        at,
        Offset(math.cos(a) * sp, math.sin(a) * sp),
        1.2 + _rng.nextDouble() * 2.4,
      ));
    }
    if (droplets.length > 120) {
      droplets.removeRange(0, droplets.length - 120);
    }
  }

  void update(double dt, double surfaceY, {double gravity = 900}) {
    for (final b in bubbles) {
      b.phase += dt * b.wobble;
      b.pos = Offset(
        b.pos.dx + math.sin(b.phase) * 12 * dt,
        b.pos.dy - b.speed * dt,
      );
      if (b.pos.dy < surfaceY) b.life = 0;
    }
    bubbles.removeWhere((b) => b.life <= 0);

    for (final d in droplets) {
      d.vel = Offset(d.vel.dx, d.vel.dy + gravity * dt);
      d.pos = d.pos + d.vel * dt;
      d.life -= dt * 1.8;
    }
    droplets.removeWhere((d) => d.life <= 0);
  }

  static double lerp(double a, double b, double t) => a + (b - a) * t;
}

File: ZeroFoamChallenge/Sources/Engine/Effects.swift
import CoreGraphics
import Foundation

struct Bubble {
    var pos: CGPoint
    var radius: CGFloat
    var speed: CGFloat
    var wobble: CGFloat
    var phase: CGFloat
    var alive = true
}

struct Droplet {
    var pos: CGPoint
    var vel: CGPoint
    var radius: CGFloat
    var life: CGFloat = 1
}

/// Purely cosmetic particle systems: fizzy bubbles inside the cola and splash
/// droplets at the impact point.
final class Effects {
    private(set) var bubbles: [Bubble] = []
    private(set) var droplets: [Droplet] = []

    private var bubbleAccumulator: CGFloat = 0
    private var rng = SystemRandomNumberGenerator()

    func clear() {
        bubbles.removeAll(keepingCapacity: true)
        droplets.removeAll(keepingCapacity: true)
    }

    private func rand() -> CGFloat { CGFloat.random(in: 0...1, using: &rng) }

    func spawnBubbles(dt: CGFloat, rate: CGFloat,
                      left: CGFloat, right: CGFloat,
                      bottomY: CGFloat, surfaceY: CGFloat) {
        guard right > left, bottomY > surfaceY else { return }
        bubbleAccumulator += rate * dt
        while bubbleAccumulator >= 1 {
            bubbleAccumulator -= 1
            let x = lerp(left, right, rand())
            let y = lerp(surfaceY, bottomY, 0.35 + rand() * 0.65)
            bubbles.append(Bubble(pos: CGPoint(x: x, y: y),
                                  radius: 1.2 + rand() * 2.6,
                                  speed: 26 + rand() * 52,
                                  wobble: 3 + rand() * 7,
                                  phase: rand() * .pi * 2))
        }
        if bubbles.count > 220 { bubbles.removeFirst(bubbles.count - 220) }
    }

    func splash(at point: CGPoint, strength: CGFloat) {
        let n = min(max(Int(strength * 3), 0), 4)
        for _ in 0..<n {
            let a = -CGFloat.pi / 2 + (rand() - 0.5) * 2.2
            let sp = 60 + rand() * 180 * strength
            droplets.append(Droplet(pos: point,
                                    vel: CGPoint(x: cos(a) * sp, y: sin(a) * sp),
                                    radius: 1.2 + rand() * 2.4))
        }
        if droplets.count > 120 { droplets.removeFirst(droplets.count - 120) }
    }

    func update(dt: CGFloat, surfaceY: CGFloat, gravity: CGFloat = 900) {
        for i in bubbles.indices {
            bubbles[i].phase += dt * bubbles[i].wobble
            bubbles[i].pos.x += sin(bubbles[i].phase) * 12 * dt
            bubbles[i].pos.y -= bubbles[i].speed * dt
            if bubbles[i].pos.y < surfaceY { bubbles[i].alive = false }
        }
        bubbles.removeAll { !$0.alive }

        for i in droplets.indices {
            droplets[i].vel.y += gravity * dt
            droplets[i].pos = droplets[i].pos + droplets[i].vel * dt
            droplets[i].life -= dt * 1.8
        }
        droplets.removeAll { $0.life <= 0 }
    }
}
